### PR TITLE
mgr/orchestrator: Add test orchestrator

### PR DESCRIPTION
**Pull request by sebastian-philipp**
1. To be able to run the cli without an external orchestrator.
2. Run the CLI in Teuthology.

Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>

### TODO
- [x] Make the call to ceph-volume actually work. See #25237
- [x] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by tchaikov** [comment]
retest this please

**Comment by sebastian-philipp** [comment]
@alfredodeza This PR now contains the [changes](https://github.com/ceph/ceph/pull/25236/commits/67fc10d317d4139d2b3472b6362a99f10bafdd85) needed to create the virtual environment as discussed in #25237

File: src/pybind/mgr/test_orchestrator/module.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+            """
+            c_v_out = check_output(cmd, shell=True)
+
+        for out in c_v_out.splitlines():
+            if not out.startswith(b'-->') and not out.startswith(b' stderr'):
+                self.log.error(out)
+                devs = []
+                for device in json.loads(out):
+                    dev = orchestrator.InventoryDevice()
+                    if 'partitions' not in device["sys_api"]:
```
Comment (sebastian-philipp) [inline src/pybind/mgr/test_orchestrator/module.py]: This filters our partitions in the JSON output. Relates to #25390. Can be removed afterwards

**Comment by sebastian-philipp** [comment]
@alfredodeza ping?

File: src/pybind/mgr/test_orchestrator/module.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        self._complete = True
+
+
+class TestWriteCompletion(orchestrator.WriteCompletion):
+    def __init__(self, execute_cb, complete_cb, message):
+        super(TestWriteCompletion, self).__init__()
+        self.execute_cb = execute_cb
+        self.complete_cb = complete_cb
+
+        # Executed means I executed my k8s API call, it may or may
```
Comment (tserong) [inline src/pybind/mgr/test_orchestrator/module.py]: Probably souldn't mention k8s

File: src/pybind/mgr/test_orchestrator/module.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+            # TODO: check self.result for API errors
+            if self.complete_cb is None:
+                self.effective = True
+            else:
+                self.effective = self.complete_cb()
+
+
+def deferred_read(f):
+    """
+    Decorator to make RookOrchestrator methods return
```
Comment (tserong) [inline src/pybind/mgr/test_orchestrator/module.py]: s/RookOrchestrator/TestOrchestrator/

File: src/pybind/mgr/test_orchestrator/module.py (hunk 1 above)
Comment (tserong) [reply src/pybind/mgr/test_orchestrator/module.py]: #25390 is merged now, so I guess this can be removed?

**Comment by sebastian-philipp** [comment]
@tserong 
> Instead of setting up a virtualenv, is there some way ceph-volume could be installed to build/bin, along with the ceph cli tool and the other binaries? Because that'd be kinda neat :-)

Actually, you don't need to install it at all. You can invoke c-v directly from the source folder. #25237 is the only thing missing here.

**Comment by tserong** [comment]
> Actually, you don't need to install it at all. You can invoke c-v directly from the source folder. #25237 is the only thing missing here.

I guess I meant: be able to invoke it from the build/bin directory, the same way you can invoke build/bin/ceph currently. Would ceph-mgr be able to find ceph-volume then, in that directory, without having to activate a virtualenv internally?

File: src/pybind/mgr/test_orchestrator/module.py (hunk 2 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/test_orchestrator/module.py]: done.

File: src/pybind/mgr/test_orchestrator/module.py (hunk 3 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/test_orchestrator/module.py]: done.

File: src/pybind/mgr/orchestrator.py [hunk 4]
```diff
@@ -151,6 +151,7 @@ def get_inventory(self, node_filter=None):
 
         :param node_filter:
         :return: list of InventoryNode
+        :rtype: InventoryNode
```
[the remark this answers is not in the document]
Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator.py]: removed form this  PR. Thanks!

**Comment by sebastian-philipp** [comment]
```
2018-12-20 10:22:17,176.176 INFO:__main__:test_status (tasks.mgr.test_orchestrator_cli.TestDashboard) ... ok
2018-12-20 10:22:17,176.176 INFO:__main__:Stopped test: test_status (tasks.mgr.test_orchestrator_cli.TestDashboard) in 9.691915s
2018-12-20 10:22:17,176.176 INFO:__main__:
2018-12-20 10:22:17,176.176 INFO:__main__:----------------------------------------------------------------------
2018-12-20 10:22:17,177.177 INFO:__main__:Ran 3 tests in 66.019s
2018-12-20 10:22:17,177.177 INFO:__main__:
2018-12-20 10:22:17,177.177 INFO:__main__:OK

```

**Comment by sebastian-philipp** [comment]
@liewegas . I just added the missing `qa/suites/rados/mgr/tasks/orchestrator_cli.yaml`.

**Pull request merged**